### ISSUE TITLE
How to implement CertificateProvider

### ISSUE THREAD

**Issue by noahdav** [grpc/grpc#25172]
### What version of gRPC and what language are you using?
1.34.1

### What operating system (Linux, Windows,...) and version?
Windows

### What runtime / compiler are you using (e.g. python version or version of gcc)
[INFO] [27] -- The C compiler identification is MSVC 19.0.24234.1
[INFO] [28] -- The CXX compiler identification is MSVC 19.0.24234.1

### What did you do?
We recently upgraded to the latest release 1.34.1.  Among other changes the TlsCredentialsOptions have changed.  Prior to this change we implemented the grpc::experimental::TlsCredentialReloadInterface and registered it as part of the grpc::experimental::TlsCredentialsOptions.  In our environment certificates rotate frequently so we need a way to refresh certificates.

It looks like the changes described https://github.com/grpc/proposal/pull/205/files#diff-ba0e9b8ef8852e7177ea0bfb1cd507ebb8e70e9c645e1ea2964c1c7e66560127R313 have made their way into this release.  Overall this seems like a good approach.  However I am confused on how to use this.

Now with the current changes it looks like we need to implement the CertificateProviderInterface.  I was trying to look at how this was done.  The examples in the above PR seem to be outdated and many of the classes/functions do not exist.  By looking at the StaticDataCertificateProvider I tried to understand what is happening here.  The problem for us is that the implementation happens in grpc_core.  This alone is not an issue but most of the work is actually done in the grpc_tls_certificate_distributor.  Since we do not have access to the grpc_core from our application I am not seeing how this is intended to work.  Are there any working examples of how applications can implement custom providers?

Since the reload is now removed I am not sure how to move forward.  Can someone please help us understand how to implement a custom provider?

**Comment by jiangtaoli2016** [comment]
@ZhenLian and @markdroth FYI.

**Comment by markdroth** [comment]
@ZhenLian, can you help this user with this?  Seeing what problems they run into will probably help us ensure that we've really thought through the new API.

Thanks!

**Comment by jiangtaoli2016** [comment]
It looks like current design does not allow external users to plug in their own cert provider.

**Comment by markdroth** [comment]
Oh, good point.  But to be clear, the *design* does allow for external provider implementations; it's just that the APIs needed to implement an external provider have not yet been implemented.

@noahdav, do you need a custom provider implementation, or can you use the file-watcher provider that we implemented?

@jiangtaoli2016 If the user really needs a custom provider implementation, we should prioritize implementing those APIs.

**Comment by noahdav** [comment]
@markdroth In our environment which runs on bare metal azure nodes we have a custom PKI infrastructure.  We have internal APIs which we can call to lookup the current cert and also verify cert.  The issue is that the cert rotates quite often so we need a way to update the certificate without tearing down the channel.  Today the certificates are stored in the windows cert store in custom stores.  Our APIs will give us the CERT_CONTEXT which we us and convert to PEM.  Going forward the PEM files will exist on disk but as of now this is not yet available to us.

We also do not currently know when the certs are rotated but have the ability to lookup the current cert which is guaranteed to be the current cert.  In the prior implementation we used the TlsCredentialReloadInterface so that we could plumb the new cert when the connection failure occurred.  Since this is no longer available I am not sure what I can do.

As for your original question can you point me towards this file-watcher provider?  I was looking at the StaticDataCertificateProvider but from what I saw it lacked the ability to reload once the cert was rotated.

**Comment by markdroth** [comment]
The `FileWatcherCertificateProvider` is here:

https://github.com/grpc/grpc/blob/0533276facb2844634a8d2ba981adefe4992bb4e/include/grpcpp/security/tls_certificate_provider.h#L91

With that, you could have an external process use your internal APIs and dump the resulting certs to disk, and then read them with the file-watcher provider.

The new provider API is specifically designed to *not* be called on every single connection attempt, because that can add a lot of overhead on busy systems.  However, provider implementations that don't get notified on changes can periodically scan for changes at whatever frequency they want.  (This is actually how the file-watcher provider itself works.)

**Comment by noahdav** [comment]
Thanks @markdroth.  We will look at this provider.  We also use an openssl engine to avoid exporting the private key.  Will this file watcher still maintain the same engine: syntax for the private key?

**Comment by markdroth** [comment]
I don't know the current state of the openssl engine integration.  @jiangtaoli2016 or @ZhenLian, can you advise?

**Comment by noahdav** [comment]
Looking more the FileWatcherCertificateProvider is not in the latest release https://github.com/grpc/grpc/releases/tag/v1.34.1.  When can we expect a stable release with this included?

You mentioned that you this is not being called on every connection attempt.  Is this going to improve for authorization too?  We currently use the AuthMetadataProcessor to authorize but this occurs on every request.  What is the recommendation for this going forward?

**Comment by jiangtaoli2016** [comment]
It has the same syntax on Engine. It would work for OpenSSL Engine automatically. Let me know if it is not the case. @noahdav

**Comment by markdroth** [comment]
Ah, sorry; I didn't realize that the file-watcher cert provider wasn't implemented by the time we cut the 1.34 release branch.  But it is already in master, so it will be in 1.35.

The `AuthMetadataProcessor` is not affected by these changes; it will still happen on every request.  That actually has nothing to do with connections or TLS at all; it authorizes individual requests (not connections) based on the metadata on the individual request.  Because each request can have a different set of metadata, each one needs to be authorized independently.

**Comment by noahdav** [comment]
Thanks for the pointers.  I will look at the file based provider.  For our service we have client APIs which were built on the previous implementation.  Our clients have a callback to provide the certs/pem so we can feed them into grpc.  Having to break this may be an issue for us.  Instead building our own provider might simplify this where we do not need to break our client API.  That being said we will look at the file based provider and see what we can do.

It seems to me that there may be was to optimize the auth processor.  For instance have the auth context have the ability to mark the context so that it will not be checked on additional requests.  Otherwise just continue to check on every request.  In many cases just checking once per connection may be sufficient for many services.

**Comment by markdroth** [comment]
If you want to authorize once per connection based on the client certificate, then that can be done via a certificate verification plugin that runs at connection-establishment time.  The current API for cert verification is fairly primative, but @ZhenLian is currently working on a new design, which will be part of the TLS creds gRFC.

The `AuthMetadataPlugin` is not intended to solve that problem; as its name implies, it's primarily intended for authorizing based on metadata on a per-request basis.

**Comment by ZhenLian** [comment]
Sorry, I was a bit lagged behind the discussions here, and thanks @markdroth and @jiangtaoli2016 for covering this. Yes, I am currently working on the new design that will allow for authorization customization on a per-connection basis. The gRFC is a bit out-of-date, but I will update it once the major API is finalized. @noahdav feel free to let me know if you have any questions/concerns. Thanks!

**Comment by noahdav** [comment]
Thanks @ZhenLian and @markdroth.  Is this certificate verification plugin available for us to now or is this something which we need to wait for?  Currently we have been using AuthMetadataPlugin as it was all we had from what we knew.  Unless we were missing something.  If this certificate verification plugin is available we could cardinally make use of it.

@ZhenLian can you please send me a like to the gRFC for this?

Back to our original issue with the certificate provider.  Since our service also has its own clients and an API which we support I do not think it is ideal to brake this if we do not need to.  Requiring them to export the keys to a file every time their certs are rotated seems a little much for each client.  Our current API supports receiving the buffers from our clients and passing them down to gRPC.  We do not take any dependency on the type of cert that we are dealing with.  Additionally different clients could have different rules for rotation which makes this a pain for all of our clients.  Ideally we would like to write a certificate provider which can handle certs in buffers (not files) like we had.  We would need to be able to rotate them appropriately as well.  Is there plans to allow for custom certificate providers where we could implement such a provider?

**Comment by ZhenLian** [comment]
> Thanks @ZhenLian and @markdroth. Is this certificate verification plugin available for us to now or is this something which we need to wait for? Currently we have been using AuthMetadataPlugin as it was all we had from what we knew. Unless we were missing something. If this certificate verification plugin is available we could cardinally make use of it.

The new certificate verification plugin is based on the current TlsServerAuthorizationCheckConfig(https://github.com/grpc/grpc/blob/e745f8cb558798f1c8b91baa42d46228a9d4a0c4/include/grpcpp/security/tls_credentials_options.h#L103), but the API will be very different to make the plugin composable. This TlsServerAuthorizationCheckConfig will be completely removed once the new comes in, but you probably can get some ideas of how it works.

> @ZhenLian can you please send me a like to the gRFC for this?

Since that plugin will also be part of TLS credential API, it was also listed in the gRFC you posted at the issue description page, under "Custom Authorization" section.

> Back to our original issue with the certificate provider. Since our service also has its own clients and an API which we support I do not think it is ideal to brake this if we do not need to. Requiring them to export the keys to a file every time their certs are rotated seems a little much for each client. Our current API supports receiving the buffers from our clients and passing them down to gRPC. We do not take any dependency on the type of cert that we are dealing with. Additionally different clients could have different rules for rotation which makes this a pain for all of our clients. Ideally we would like to write a certificate provider which can handle certs in buffers (not files) like we had. We would need to be able to rotate them appropriately as well. Is there plans to allow for custom certificate providers where we could implement such a provider?

Basically, I think something like a in-memory reloading provider would be suitable for your case, right? Implementing an external provider(which is not supported right now but will be supported in the future) would definitely work, but I am thinking if it is necessary to have an internal in-memory provider(something like InMemoryCertWatcherCertificateProvider), for easy use. We currently have an internal request asking for the similar in-memory reloading provider. I guess if this pattern is general enough, we might also make it an internal implementation. Can I know your opinions based on that? @markdroth @jiangtaoli2016  Thanks!

**Comment by jiangtaoli2016** [comment]
IIUC, we just need a certificate provider where there are two APIs called `UpdateIdentityCertificates()` and `UpdateRootCertificates()` to allow applications to call these two functions whenever they want to update certificates. That should be straight-forward.

**Comment by markdroth** [comment]
I think it could be useful to have a type of provider that has methods to allow the application to set the certs on an ad-hoc basis without having to actually implement a provider plugin.  This could basically be similar to the static provider, except that (a) it would need to support multiple cert names and (b) it would need to be able to accept new data after instantiation time.  Or you could think of it as being similar to the file-watcher provider, but instead of watching files, it gets new data from the methods invoked by the application.

That having been said, it's not clear to me that such a provider implementation would actually be helpful in this particular use-case.  It sounds like @noahdav wants to implement a custom provider anyway, so I think what we really need here is just to expose the external provider API.

**Comment by noahdav** [comment]
@jiangtaoli2016 that sounds pretty straight forward.  I think that would work in our case.

**Comment by noahdav** [comment]
@markdroth Is there any update on this?  I am still needing to implement an in memory way to update the certs.  We are currently on 1.36.0 due to the certificate issues.  In order to update we need a way to update the existing certs.  If this is implementing a new provider or the plugin you suggested I am fine with either way.  My understanding is that the APIs to do this are not exposed to clients.  Is there a plan to expose this so we can create a provider/plugin for this?

**Comment by markdroth** [comment]
@ZhenLian, what is the ETA for the external credential provider API?

**Comment by ZhenLian** [comment]
We haven't got a chance to plan this out yet - @yihuazhang, would you please be able to help?
Mark, what do you think about the idea to directly have an in-memory provider, similar to FileWatcherCertificateProvider? As you pointed out in a previous comment, we can think of the in-memory provider as a file-watcher provider which gets new data from the methods invoked by the applications, so the implementation can be fairly straight-forward. Plus, this case is common enough - we've got some internal inquires asking about the exact same thing. WDYT?

**Comment by markdroth** [comment]
Yes, I think that's a good idea.

**Comment by yihuazhang** [comment]
Hi @noahdav, Is it possible for you to implement @ZhenLian's idea mentioned in https://github.com/grpc/grpc/issues/25172#issuecomment-861107452?  It seems pretty straightforward and we can provide guidance and review on your PR. Zhen is pretty tied up in Q3.

**Comment by noahdav** [comment]
@yihuazhang I am pretty booked now but will be getting to this in the next month or so.  I can probably pick this up then.  If you can help with design prior to that this would be helpful.

**Comment by yihuazhang** [comment]
Sure. That's reasonable. @ZhenLian could you please provide a detailed design expanding your idea in https://github.com/grpc/grpc/issues/25172#issuecomment-861107452?

**Comment by ZhenLian** [comment]
As discussed offline with @yihuazhang, we are planning to assign this to one of our interns tentatively. I will update on this thread when we have more context on how this is going to be addressed. Thanks!

**Comment by noahdav** [comment]
@ZhenLian Thanks for the update.  This works even better for us.  Please let me know if you need additional help with this.

**Comment by itsemmanuelfrancis** [comment]
PR to fix this issue: https://github.com/grpc/grpc/pull/26663
Example code:
```
  std::string test_root_certificate = "initial_root_certificate";
  IdentityKeyCertPair identity_pair = {
      "initial_private_key",        // private_key
      "initial_certificate_chain",  // certificate_chain
  };
  auto identity_pairs = std::vector<IdentityKeyCertPair>(1, identity_pair);
  auto certificate_provider = std::make_shared<DataWatcherCertificateProvider>(
      root_certificate, identity_pairs);
  // ... at some point when we need reloading
  test_root_certificate = "updated_root_certificate";
  grpc::Status reload_root_status =
      certificate_provider->ReloadRootCertificate(test_root_certificate);
  identity_pair = {
      "updated_private_key",        // private_key
      "updated_certificate_chain",  // certificate_chain
  };
  identity_pairs.assign(1, identity_pair);
  grpc::Status reload_key_cert_pair_status =
      certificate_provider->ReloadKeyCertificatePair(identity_pairs);
```
FYI: @ZhenLian

**Comment by ZhenLian** [comment]
@noahdav Can you please take a look at the code snippet on the new `DataWatcherCertificateProvider` API and see if that meets your needs? Thank you!

**Comment by noahdav** [comment]
This look like it should work for us.  However one improvement that would be nice to have a reverse flow so that we could signal the provider that new certs are available.  This way the provider could call into the application asking for the new certs.  This would simplify that the initial and updated certs both are retrieved in the same way.

**Comment by ZhenLian** [comment]
Thanks for the feedback!
IIUC, you are referring to the old-fashioned way of credential reloading, where user specifies a callback and we just invoke that callback every time when there is a new connection, right? 
We changed the whole reloading flow to be users actively telling gRPC when to reload certs, for reasons like performance, API simplicity, etc. If we want that reversed flow, we will likely need the `external provider`.
Just to make sure, Does `DataWatcherCertificateProvider` fit all our needs here? If so, we will go ahead and implement it. The  `external provider` will be deemed as an advanced feature and might be added later on.

**Comment by itsemmanuelfrancis** [comment]
^^ @noahdav

**Comment by noahdav** [comment]
sorry for the delay.  This looks like it will meet our needs.

@ZhenLian thanks for the context.  I was thinking that the user could tell the provider that new creds are available and the provider could then call the callback but this really does not make much sense anymore.  If we know that creds are available then we can just update the provider as you specified in the DataWatcherCertificateProvider.  The external provider may solve this but it is a nice to have not something that is required.

Thanks for getting to this feature.  It will be very useful for us and I assume others too.

**Comment by markdroth** [comment]
It sounds like this is now resolved.  If there's more to do here, please comment, and we can reopen.

**Comment by immiao** [comment]
@markdroth is there a C# implementation for the file watcher cert provider?
